src/PrimeTower.cpp

```diff
@@ -97,23 +102,62 @@ void PrimeTower::generateGroundpoly()
 
 void PrimeTower::generatePaths(const SliceDataStorage& storage)
 {
-    would_have_actual_tower = storage.max_print_height_second_to_last_extruder >= 0; //Maybe it turns out that we don't need a prime tower after all because there are no layer switches.
+    would_have_actual_tower
+        = storage.max_print_height_second_to_last_extruder >= 0; // Maybe it turns out that we don't need a prime tower after all because there are no layer switches.
     if (would_have_actual_tower && enabled)
     {
         generatePaths_denseInfill();
         generateStartLocations();
     }
 }
 
+PrimeTower::ExtrusionMoves PrimeTower::generatePaths_base(const Polygons& inset, size_t rings, coord_t line_width)
+{
+    ExtrusionMoves pattern;
+
+    Polygons path = inset.offset(line_width / 2);
+    for (size_t ring = 0; ring < rings; ++ring)
+    {
+        pattern.polygons.add(path);
+        path = path.offset(line_width);
+    }
+
+    return pattern;
+}
+
+PrimeTower::ExtrusionMoves PrimeTower::generatePaths_inset(const Polygons& outer_poly, coord_t line_width, coord_t initial_inset)
+{
+    const Scene& scene = Application::getInstance().current_slice->scene;
+    const Settings& mesh_group_settings = scene.current_mesh_group->settings;
+
+    ExtrusionMoves pattern;
+
+    Polygons inset = outer_poly.offset(-(initial_inset + line_width / 2));
+    while (! inset.empty())
+    {
+        pattern.polygons.add(inset);
+        inset = inset.offset(-line_width);
+    }
+
+    return pattern;
+}
+
 void PrimeTower::generatePaths_denseInfill()
 {
     const Scene& scene = Application::getInstance().current_slice->scene;
     const Settings& mesh_group_settings = scene.current_mesh_group->settings;
     const coord_t layer_height = mesh_group_settings.get<coord_t>("layer_height");
+    const bool base_enabled = mesh_group_settings.get<bool>("prime_tower_brim_enable");
```

Comment:
out of scope but not a fan of magic strings

Reply:
Me neither :disappointed: but this is how the whole engine is implemented until now...